File: ios/ShouldIGo/Features/Conditions/BeachSketchView.swift
import SwiftUI

// MARK: - Direction lookup

private let directionDegrees: [String: Double] = [
    "N": 0, "NNE": 22.5, "NE": 45, "ENE": 67.5,
    "E": 90, "ESE": 112.5, "SE": 135, "SSE": 157.5,
    "S": 180, "SSW": 202.5, "SW": 225, "WSW": 247.5,
    "W": 270, "WNW": 292.5, "NW": 315, "NNW": 337.5,
]

private func compassToVec(_ deg: Double) -> CGPoint {
    let rad = (deg - 90) * .pi / 180
    return CGPoint(x: cos(rad), y: sin(rad))
}

private func windRelation(waveDeg: Double, windDeg: Double) -> String {
    let diff = abs(((waveDeg - windDeg + 540).truncatingRemainder(dividingBy: 360)) - 180)
    if diff < 50 { return "Onshore" }
    if diff > 130 { return "Offshore" }
    return "Cross-shore"
}

// MARK: - Beach Sketch View

struct BeachSketchView: View {
    let waveDirection: String?
    let windDirection: String?

    private let W: CGFloat = 400
    private let H: CGFloat = 220

    private var waveDeg: Double? {
        guard let d = waveDirection else { return nil }
        return directionDegrees[d]
    }
    private var windDeg: Double? {
        guard let d = windDirection else { return nil }
        return directionDegrees[d]
    }

    var body: some View {
        if let wd = waveDeg {
            VStack(alignment: .leading, spacing: 10) {
                SectionHeader("Beach View")

                Canvas { context, size in
                    drawSketch(context: context, size: size, waveDeg: wd)
                }
                .frame(height: 180)
                .clipShape(RoundedRectangle(cornerRadius: 12))

                // Footer labels
                HStack(spacing: 16) {
                    if let dir = waveDirection {
                        Label("Swell \(dir)", systemImage: "water.waves")
                            .font(.caption.weight(.semibold))
                            .foregroundStyle(Color(red: 0.15, green: 0.39, blue: 0.93)) // #2563eb
                    }
                    if let wdeg = windDeg, let wvdeg = waveDeg {
                        let relation = windRelation(waveDeg: wvdeg, windDeg: wdeg)
                        Label("\(relation)\(windDirection.map { " (\($0))" } ?? "")",
                              systemImage: "wind")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }
                }
            }
            .padding(16)
            .background(Color("cardBackground"), in: RoundedRectangle(cornerRadius: 20))
        }
    }

    // MARK: - Drawing

    private func drawSketch(context: GraphicsContext, size: CGSize, waveDeg: Double) {
        let scaleX = size.width / W
        let scaleY = size.height / H

        var ctx = context
        ctx.scaleBy(x: scaleX, y: scaleY)

        // Background water
        ctx.fill(Path(roundedRect: CGRect(x: 0, y: 0, width: W, height: H), cornerRadius: 6),
                 with: .color(Color(red: 0.91, green: 0.96, blue: 0.99).opacity(0.6)))

        let seaVec = compassToVec(waveDeg)
        let inlandVec = CGPoint(x: -seaVec.x, y: -seaVec.y)
        let mid = CGPoint(x: W / 2, y: H / 2)

        // Shore angle (perpendicular to wave direction)
        let shoreAngle = atan2(-seaVec.x, seaVec.y)
        let shorePv = CGPoint(x: cos(shoreAngle), y: sin(shoreAngle))

        // 1. Bathymetry contour arcs
        let contours: [(dist: CGFloat, halfLen: CGFloat, opacity: Double, width: CGFloat)] = [
            (32, 105, 0.4, 1.2),
            (70, 120, 0.28, 0.9),
            (115, 108, 0.18, 0.7),
        ]
        for c in contours {
            let arcPath = shoreParallelArc(mid: mid, seaVec: seaVec, dist: c.dist,
                                           halfLen: c.halfLen, shoreAngle: shoreAngle)
            ctx.stroke(arcPath, with: .color(Color(red: 0.38, green: 0.65, blue: 0.98).opacity(c.opacity)),
                       lineWidth: c.width)
        }

        // 2. Beach shape (simple arc)
        let beachPath = makeBeachPath(mid: mid, inlandVec: inlandVec, shorePv: shorePv)
        ctx.fill(beachPath, with: .color(Color(red: 0.90, green: 0.83, blue: 0.60).opacity(0.7)))
        ctx.stroke(beachPath, with: .color(Color(red: 0.54, green: 0.46, blue: 0.38)), lineWidth: 1.5)

        // 3. Foam / shore-break
        let foamPath = makeFoamPath(mid: mid, seaVec: seaVec, shoreAngle: shoreAngle)
        ctx.stroke(foamPath, with: .color(Color(red: 0.75, green: 0.86, blue: 0.99).opacity(0.7)),
                   style: StrokeStyle(lineWidth: 4, lineCap: .round))

        // 4. Wave crest lines
        let waveLines: [(dist: CGFloat, halfLen: CGFloat, opacity: Double, width: CGFloat)] = [
            (15, 55, 0.55, 1.8),
            (50, 68, 0.45, 1.5),
            (88, 72, 0.35, 1.2),
            (130, 65, 0.25, 1.0),
            (170, 55, 0.18, 0.8),
        ]
        for w in waveLines {
            let wavePath = shoreParallelArc(mid: mid, seaVec: seaVec, dist: w.dist,
                                            halfLen: w.halfLen, shoreAngle: shoreAngle)
            ctx.stroke(wavePath, with: .color(Color.blue.opacity(w.opacity)), lineWidth: w.width)
        }

        // 5. Swell direction arrow (blue)
        let maxDist = maxSeawardDist(mid: mid, vec: seaVec)
        let wvCx = mid.x + seaVec.x * maxDist * 0.70 + shorePv.x * 30
        let wvCy = mid.y + seaVec.y * maxDist * 0.70 + shorePv.y * 30
        drawArrow(ctx: &ctx, cx: wvCx, cy: wvCy, direction: inlandVec, length: 44,
                  color: Color(red: 0.15, green: 0.39, blue: 0.93), lineWidth: 2.0)
        drawLabel(ctx: &ctx, x: wvCx + seaVec.x * 26, y: wvCy + seaVec.y * 26,
                  text: "SWELL", color: Color(red: 0.15, green: 0.39, blue: 0.93))

        // 6. Wind direction arrow (amber, dashed)
        if let windDeg {
            let windTV = compassToVec((windDeg + 180).truncatingRemainder(dividingBy: 360))
            let wdCx = mid.x + seaVec.x * maxDist * 0.55 + shorePv.x * (-35)
            let wdCy = mid.y + seaVec.y * maxDist * 0.55 + shorePv.y * (-35)
            drawArrow(ctx: &ctx, cx: wdCx, cy: wdCy, direction: windTV, length: 36,
                      color: Color(red: 0.96, green: 0.62, blue: 0.04), lineWidth: 1.5,
                      dashed: true)
            drawLabel(ctx: &ctx, x: wdCx - windTV.x * 22, y: wdCy - windTV.y * 22,
                      text: "WIND", color: Color(red: 0.96, green: 0.62, blue: 0.04))
        }

        // 7. Compass rose (top-right)
        let compassX = W - 16
        drawLabel(ctx: &ctx, x: compassX, y: 12, text: "N", color: Color.gray)
        var arrowPath = Path()
        arrowPath.move(to: CGPoint(x: compassX, y: 18))
        arrowPath.addLine(to: CGPoint(x: compassX, y: 27))
        ctx.stroke(arrowPath, with: .color(.gray.opacity(0.6)), lineWidth: 1.2)
        var tri = Path()
        tri.move(to: CGPoint(x: compassX, y: 18))
        tri.addLine(to: CGPoint(x: compassX - 3, y: 24))
        tri.addLine(to: CGPoint(x: compassX + 3, y: 24))
        tri.closeSubpath()
        ctx.fill(tri, with: .color(.gray.opacity(0.6)))
    }

    // MARK: - Drawing helpers

    private func shoreParallelArc(mid: CGPoint, seaVec: CGPoint, dist: CGFloat,
                                   halfLen: CGFloat, shoreAngle: CGFloat) -> Path {
        let cx = mid.x + seaVec.x * dist
        let cy = mid.y + seaVec.y * dist
        let pv = CGPoint(x: cos(shoreAngle), y: sin(shoreAngle))
        let p1 = CGPoint(x: cx - pv.x * halfLen, y: cy - pv.y * halfLen)
        let p2 = CGPoint(x: cx + pv.x * halfLen, y: cy + pv.y * halfLen)
        let cp = CGPoint(x: cx + seaVec.x * 8, y: cy + seaVec.y * 8)
        var path = Path()
        path.move(to: p1)
        path.addQuadCurve(to: p2, control: cp)
        return path
    }

    private func makeBeachPath(mid: CGPoint, inlandVec: CGPoint, shorePv: CGPoint) -> Path {
        let cx = mid.x + inlandVec.x * 35
        let cy = mid.y + inlandVec.y * 35
        let halfLen: CGFloat = 130
        let width: CGFloat = 22

        let p1 = CGPoint(x: cx - shorePv.x * halfLen, y: cy - shorePv.y * halfLen)
        let p2 = CGPoint(x: cx + shorePv.x * halfLen, y: cy + shorePv.y * halfLen)
        let p3 = CGPoint(x: p2.x + inlandVec.x * width, y: p2.y + inlandVec.y * width)
        let p4 = CGPoint(x: p1.x + inlandVec.x * width, y: p1.y + inlandVec.y * width)

        var path = Path()
        path.move(to: p1)
        path.addLine(to: p2)
        path.addLine(to: p3)
        path.addLine(to: p4)
        path.closeSubpath()
        return path
    }

    private func makeFoamPath(mid: CGPoint, seaVec: CGPoint, shoreAngle: CGFloat) -> Path {
        shoreParallelArc(mid: mid, seaVec: seaVec, dist: 5, halfLen: 120, shoreAngle: shoreAngle)
    }

    private func maxSeawardDist(mid: CGPoint, vec: CGPoint) -> CGFloat {
        let margin: CGFloat = 18
        let bounds = [
            vec.x > 0.01 ? (W - margin - mid.x) / vec.x : CGFloat.infinity,
            vec.x < -0.01 ? (margin - mid.x) / vec.x : CGFloat.infinity,
            vec.y > 0.01 ? (H - margin - mid.y) / vec.y : CGFloat.infinity,
            vec.y < -0.01 ? (margin - mid.y) / vec.y : CGFloat.infinity,
        ]
        return min(160, bounds.filter { $0 > 0 }.min() ?? 160)
    }

    private func drawArrow(ctx: inout GraphicsContext, cx: CGFloat, cy: CGFloat,
                           direction: CGPoint, length: CGFloat, color: Color,
                           lineWidth: CGFloat, dashed: Bool = false) {
        let halfLen = length / 2
        let s = CGPoint(x: cx - direction.x * halfLen, y: cy - direction.y * halfLen)
        let e = CGPoint(x: cx + direction.x * halfLen, y: cy + direction.y * halfLen)

        var line = Path()
        line.move(to: s)
        line.addLine(to: e)
        let style = dashed
            ? StrokeStyle(lineWidth: lineWidth, dash: [5, 3])
            : StrokeStyle(lineWidth: lineWidth)
        ctx.stroke(line, with: .color(color), style: style)

        // Arrowhead
        let hw: CGFloat = 5, hl: CGFloat = 9
        let pv = CGPoint(x: -direction.y, y: direction.x)
        var head = Path()
        head.move(to: e)
        head.addLine(to: CGPoint(x: e.x - direction.x * hl + pv.x * hw,
                                 y: e.y - direction.y * hl + pv.y * hw))
        head.addLine(to: CGPoint(x: e.x - direction.x * hl - pv.x * hw,
                                 y: e.y - direction.y * hl - pv.y * hw))
        head.closeSubpath()
        ctx.fill(head, with: .color(color))
    }

    private func drawLabel(ctx: inout GraphicsContext, x: CGFloat, y: CGFloat,
                           text: String, color: Color) {
        let resolved = ctx.resolve(Text(text)
            .font(.system(size: 7.5, weight: .semibold, design: .serif))
            .foregroundColor(color))
        ctx.draw(resolved, at: CGPoint(x: x, y: y), anchor: .center)
    }
}

File: android/app/src/main/java/surf/shouldigo/app/feature/conditions/components/BeachSketchView.kt
package surf.shouldigo.app.feature.conditions.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.toArgb
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.nativeCanvas
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import surf.shouldigo.app.ui.theme.*
import kotlin.math.*

private val directionDegrees = mapOf(
    "N" to 0.0, "NNE" to 22.5, "NE" to 45.0, "ENE" to 67.5,
    "E" to 90.0, "ESE" to 112.5, "SE" to 135.0, "SSE" to 157.5,
    "S" to 180.0, "SSW" to 202.5, "SW" to 225.0, "WSW" to 247.5,
    "W" to 270.0, "WNW" to 292.5, "NW" to 315.0, "NNW" to 337.5
)

private fun compassToVec(deg: Double): Offset {
    val rad = Math.toRadians(deg - 90)
    return Offset(cos(rad).toFloat(), sin(rad).toFloat())
}

private fun windRelation(waveDeg: Double, windDeg: Double): String {
    val diff = abs(((waveDeg - windDeg + 540) % 360) - 180)
    return when {
        diff < 50 -> "Onshore"
        diff > 130 -> "Offshore"
        else -> "Cross-shore"
    }
}

@Composable
fun BeachSketchView(waveDirection: String?, windDirection: String?) {
    val waveDeg = waveDirection?.let { directionDegrees[it] } ?: return
    val windDeg = windDirection?.let { directionDegrees[it] }

    Surface(
        shape = RoundedCornerShape(20.dp),
        color = CardBackground
    ) {
        Column(modifier = Modifier.padding(16.dp)) {
            SectionHeader("Beach View")
            Spacer(modifier = Modifier.height(10.dp))

            Canvas(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(180.dp)
            ) {
                drawSketch(waveDeg, windDeg)
            }

            Spacer(modifier = Modifier.height(10.dp))
            Row(horizontalArrangement = Arrangement.spacedBy(16.dp)) {
                Text(
                    text = "Swell $waveDirection",
                    fontSize = 12.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = Color(0xFF2563EB)
                )
                if (windDeg != null) {
                    val relation = windRelation(waveDeg, windDeg)
                    Text(
                        text = "$relation${windDirection?.let { " ($it)" } ?: ""}",
                        fontSize = 12.sp,
                        color = SecondaryText
                    )
                }
            }
        }
    }
}

private fun DrawScope.drawSketch(waveDeg: Double, windDeg: Double?) {
    val W = size.width
    val H = size.height
    val mid = Offset(W / 2, H / 2)
    val seaVec = compassToVec(waveDeg)
    val inlandVec = Offset(-seaVec.x, -seaVec.y)
    val shoreAngle = atan2(-seaVec.x.toDouble(), seaVec.y.toDouble()).toFloat()
    val shorePv = Offset(cos(shoreAngle), sin(shoreAngle))

    // Background water
    drawRect(Color(0xFFE8F5FE).copy(alpha = 0.6f))

    // Bathymetry contours
    listOf(
        Triple(32f, 105f, 0.4f),
        Triple(70f, 120f, 0.28f),
        Triple(115f, 108f, 0.18f)
    ).forEach { (dist, halfLen, opacity) ->
        drawShoreArc(mid, seaVec, dist, halfLen, shoreAngle, Color(0xFF61A5FA).copy(alpha = opacity), 1f)
    }

    // Beach shape
    val beachPath = makeBeachPath(mid, inlandVec, shorePv)
    drawPath(beachPath, Color(0xFFE6D499).copy(alpha = 0.7f))
    drawPath(beachPath, Color(0xFF8A7661), style = Stroke(width = 1.5f))

    // Foam line
    drawShoreArc(mid, seaVec, 5f, 120f, shoreAngle, Color(0xFFBFDBFF).copy(alpha = 0.7f), 4f, StrokeCap.Round)

    // Wave crest lines
    listOf(
        Triple(15f, 55f, 0.55f),
        Triple(50f, 68f, 0.45f),
        Triple(88f, 72f, 0.35f),
        Triple(130f, 65f, 0.25f),
        Triple(170f, 55f, 0.18f)
    ).forEach { (dist, halfLen, opacity) ->
        drawShoreArc(mid, seaVec, dist, halfLen, shoreAngle, Color.Blue.copy(alpha = opacity), 1.5f)
    }

    // Swell arrow (blue)
    val maxDist = maxSeawardDist(mid, seaVec, W, H)
    val wvCx = mid.x + seaVec.x * maxDist * 0.70f + shorePv.x * 30f
    val wvCy = mid.y + seaVec.y * maxDist * 0.70f + shorePv.y * 30f
    drawArrow(Offset(wvCx, wvCy), inlandVec, 44f, Color(0xFF2563EB), 2f)
    drawSmallLabel(wvCx + seaVec.x * 26f, wvCy + seaVec.y * 26f, "SWELL", Color(0xFF2563EB))

    // Wind arrow (amber, dashed)
    if (windDeg != null) {
        val windTV = compassToVec((windDeg + 180) % 360)
        val wdCx = mid.x + seaVec.x * maxDist * 0.55f + shorePv.x * (-35f)
        val wdCy = mid.y + seaVec.y * maxDist * 0.55f + shorePv.y * (-35f)
        drawArrow(Offset(wdCx, wdCy), windTV, 36f, Color(0xFFF59E0A), 1.5f, dashed = true)
        drawSmallLabel(wdCx - windTV.x * 22f, wdCy - windTV.y * 22f, "WIND", Color(0xFFF59E0A))
    }

    // Compass rose (top right)
    drawSmallLabel(W - 16f, 12f, "N", Color.Gray)
    drawLine(Color.Gray.copy(alpha = 0.6f), Offset(W - 16f, 18f), Offset(W - 16f, 27f), strokeWidth = 1.2f)
    val triPath = Path().apply {
        moveTo(W - 16f, 18f)
        lineTo(W - 19f, 24f)
        lineTo(W - 13f, 24f)
        close()
    }
    drawPath(triPath, Color.Gray.copy(alpha = 0.6f))
}

private fun DrawScope.drawShoreArc(
    mid: Offset, seaVec: Offset, dist: Float, halfLen: Float,
    shoreAngle: Float, color: Color, width: Float,
    cap: StrokeCap = StrokeCap.Butt
) {
    val cx = mid.x + seaVec.x * dist
    val cy = mid.y + seaVec.y * dist
    val pv = Offset(cos(shoreAngle), sin(shoreAngle))
    val p1 = Offset(cx - pv.x * halfLen, cy - pv.y * halfLen)
    val p2 = Offset(cx + pv.x * halfLen, cy + pv.y * halfLen)
    val cp = Offset(cx + seaVec.x * 8f, cy + seaVec.y * 8f)

    val path = Path().apply {
        moveTo(p1.x, p1.y)
        quadraticBezierTo(cp.x, cp.y, p2.x, p2.y)
    }
    drawPath(path, color, style = Stroke(width = width, cap = cap))
}

private fun makeBeachPath(mid: Offset, inlandVec: Offset, shorePv: Offset): Path {
    val cx = mid.x + inlandVec.x * 35f
    val cy = mid.y + inlandVec.y * 35f
    val halfLen = 130f
    val width = 22f

    val p1 = Offset(cx - shorePv.x * halfLen, cy - shorePv.y * halfLen)
    val p2 = Offset(cx + shorePv.x * halfLen, cy + shorePv.y * halfLen)
    val p3 = Offset(p2.x + inlandVec.x * width, p2.y + inlandVec.y * width)
    val p4 = Offset(p1.x + inlandVec.x * width, p1.y + inlandVec.y * width)

    return Path().apply {
        moveTo(p1.x, p1.y)
        lineTo(p2.x, p2.y)
        lineTo(p3.x, p3.y)
        lineTo(p4.x, p4.y)
        close()
    }
}

private fun maxSeawardDist(mid: Offset, vec: Offset, W: Float, H: Float): Float {
    val margin = 18f
    val bounds = listOfNotNull(
        if (vec.x > 0.01f) (W - margin - mid.x) / vec.x else null,
        if (vec.x < -0.01f) (margin - mid.x) / vec.x else null,
        if (vec.y > 0.01f) (H - margin - mid.y) / vec.y else null,
        if (vec.y < -0.01f) (margin - mid.y) / vec.y else null
    ).filter { it > 0 }
    return min(160f, bounds.minOrNull() ?: 160f)
}

private fun DrawScope.drawArrow(
    center: Offset, direction: Offset, length: Float,
    color: Color, lineWidth: Float, dashed: Boolean = false
) {
    val halfLen = length / 2
    val s = Offset(center.x - direction.x * halfLen, center.y - direction.y * halfLen)
    val e = Offset(center.x + direction.x * halfLen, center.y + direction.y * halfLen)

    val effect = if (dashed) PathEffect.dashPathEffect(floatArrayOf(5f, 3f)) else null
    drawLine(color, s, e, strokeWidth = lineWidth, pathEffect = effect)

    // Arrowhead
    val hw = 5f; val hl = 9f
    val pv = Offset(-direction.y, direction.x)
    val head = Path().apply {
        moveTo(e.x, e.y)
        lineTo(e.x - direction.x * hl + pv.x * hw, e.y - direction.y * hl + pv.y * hw)
        lineTo(e.x - direction.x * hl - pv.x * hw, e.y - direction.y * hl - pv.y * hw)
        close()
    }
    drawPath(head, color)
}

private fun DrawScope.drawSmallLabel(x: Float, y: Float, text: String, color: Color) {
    drawContext.canvas.nativeCanvas.drawText(
        text,
        x,
        y + 4f,
        android.graphics.Paint().apply {
            this.color = color.toArgb()
            textSize = 20f
            textAlign = android.graphics.Paint.Align.CENTER
            typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.BOLD)
            isAntiAlias = true
        }
    )
}
